contracts/TokenManager.sol

```diff
@@ -189,6 +273,262 @@ contract TokenManager is System, IApplication {
+      else if (idx == 5) {
+        mirrorAckPackage.errorCode  = uint8(iter.next().toUint());
+        success = true;
+      }
+      else break;
+      idx++;
+    }
+    return (mirrorAckPackage, success);
+  }
+
+  function mirror(address bep20Addr, uint64 expireTime) payable public returns (bool) {
+    require(ITokenHub(TOKEN_HUB_ADDR).getBep2SymbolByContractAddr(bep20Addr) == bytes32(0x00), "already bound");
+    require(!mirrorPendingRecord[bep20Addr], "mirror pending");
+    uint256 miniRelayFee = ITokenHub(TOKEN_HUB_ADDR).getMiniRelayFee();
+    require(msg.value%TEN_DECIMALS == 0 && msg.value>=mirrorFee.add(miniRelayFee), "msg.value must be N * 1e10 and greater than sum of miniRelayFee and mirrorFee");
+    require(expireTime>=block.timestamp + 120 && expireTime <= block.timestamp + 86400, "expireTime must be two minutes later and one day earlier");
+    uint8 decimals = IBEP20(bep20Addr).decimals();
+    uint256 totalSupply = IBEP20(bep20Addr).totalSupply();
+    require(convertToBep2Amount(totalSupply, decimals) <= MAX_BEP2_TOTAL_SUPPLY, "too large total supply");
+    string memory name = IBEP20(bep20Addr).name();
+    bytes memory nameBytes = bytes(name);
+    require(nameBytes.length>=1 && nameBytes.length<=32, "name length must be in [1,32]");
+    string memory symbol = IBEP20(bep20Addr).symbol();
+    bytes memory symbolBytes = bytes(symbol);
+    require(symbolBytes.length>=MINIMUM_BEP20_SYMBOL_LEN && symbolBytes.length<=MAXIMUM_BEP20_SYMBOL_LEN, "symbol length must be in [3,8]");
+    for (uint8 i = 0; i < symbolBytes.length; i++) {
+      require((symbolBytes[i]>='a' && symbolBytes[i]<='a') || (symbolBytes[i]>='A' && symbolBytes[i]<='Z'), "symbol must not contain non-alphabet");
+    }
+    address(uint160(TOKEN_HUB_ADDR)).transfer(msg.value.sub(mirrorFee));
+    mirrorPendingRecord[bep20Addr] = true;
+    bytes32 bytes32Name;
+    assembly {
+      bytes32Name := mload(add(name, 32))
+    }
+    bytes32 bytes32Symbol;
+    assembly {
+      bytes32Symbol := mload(add(symbol, 32))
+    }
+    MirrorSynPackage memory mirrorSynPackage = MirrorSynPackage({
+      mirrorSender:  msg.sender,
+      bep20Addr:     bep20Addr,
+      bep20Name:     bytes32Name,
+      bep20Symbol:   bytes32Symbol,
+      bep20Supply:   totalSupply,
+      bep20Decimals: decimals,
+      mirrorFee:     mirrorFee/TEN_DECIMALS,
+      expireTime:    expireTime
+      });
+    ICrossChain(CROSS_CHAIN_CONTRACT_ADDR).sendSynPackage(MIRROR_CHANNELID, encodeMirrorSynPackage(mirrorSynPackage), msg.value.sub(mirrorFee).div(TEN_DECIMALS));
+    return true;
+  }
+
+  function handleMirrorAckPackage(bytes memory msgBytes) internal {
+    (MirrorAckPackage memory mirrorAckPackage, bool decodeSuccess) = decodeMirrorAckPackage(msgBytes);
+    require(decodeSuccess, "unrecognized package");
+    mirrorPendingRecord[mirrorAckPackage.bep20Addr] = false;
+    if (mirrorAckPackage.errorCode == MIRROR_STATUS_SUCCESS ) {
```

Comment:
In `System.sol`, there is a code named `CODE_OK`, can we reuse that?

Reply:
Good idea.